File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/least-consecutive-cards-match/code/python/least_consecutive_cards_match.py
# ============================================================
# Least Consecutive Cards to Match
# ============================================================

from typing import List

# ============================================================
# Solution 1: Brute Force
# Time: O(N²) | Space: O(1)
# ============================================================
class Solution1:
    def minimumCardPickup(self, cards: List[int]) -> int:
        min_len = float('inf')
        for i in range(len(cards)):
            for j in range(i + 1, len(cards)):
                if cards[i] == cards[j]:
                    min_len = min(min_len, j - i + 1)
                    break
        return -1 if min_len == float('inf') else min_len

# ============================================================
# Solution 2: Hash Map (Optimal)
# Time: O(N) | Space: O(N)
# ============================================================
class Solution2:
    def minimumCardPickup(self, cards: List[int]) -> int:
        last_seen = {}
        min_len = float('inf')

        for i, card in enumerate(cards):
            if card in last_seen:
                min_len = min(min_len, i - last_seen[card] + 1)
            last_seen[card] = i

        return -1 if min_len == float('inf') else min_len

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/least-consecutive-cards-match/code/java/LeastConsecutiveCardsMatch.java
// ============================================================
// Least Consecutive Cards to Match
// ============================================================

import java.util.*;

// ============================================================
// Solution 1: Brute Force
// Time: O(N²) | Space: O(1)
// ============================================================
class Solution1 {
    public int minimumCardPickup(int[] cards) {
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < cards.length; i++)
            for (int j = i+1; j < cards.length; j++)
                if (cards[i] == cards[j]) { min = Math.min(min, j-i+1); break; }
        return min == Integer.MAX_VALUE ? -1 : min;
    }
}

// ============================================================
// Solution 2: Hash Map (Optimal)
// Time: O(N) | Space: O(N)
// ============================================================
class Solution2 {
    public int minimumCardPickup(int[] cards) {
        Map<Integer, Integer> last = new HashMap<>();
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < cards.length; i++) {
            if (last.containsKey(cards[i]))
                min = Math.min(min, i - last.get(cards[i]) + 1);
            last.put(cards[i], i);
        }
        return min == Integer.MAX_VALUE ? -1 : min;
    }
}
